training/heterogeneous-clusters/pt.grpc.local/README.md

```diff
@@ -0,0 +1,106 @@
+70: avg step time: 2.5143303409714335
+80: avg step time: 2.530583710625001
+90: avg step time: 2.5414934969444403
+100: avg step time: 2.5489751799700024
+Training completed!
+```    
+    
+    
+B. Split data pre-processing and training testing using gRPC Client-Server inter-process communication
+---
+In this example, we are decoupling the data pre-processing component of our training job, and the deep neural network (DNN) code. This way the data processing can run on CPU instance, and DNN on GPU instance. Here by introducing concept of heterogenous instances, but demonstrated both these processes running locally. The inter-process communucation is implemented by gRPC Client-Server communication. 
+
+**Step 1**: Run gRPC Server in a new terminal session. The set of processes wait for client to request data. On request, it read the data from `data` folder, preprocess the data, and send it to the client for training. 
+     
+```
+python main_grpc_server.py --batch-size 8192 --num-data-workers 4 --iterations 100 --grpc-workers 2
+```
+where,  
+`batch-size` any integer  
+`num-workers` based on no. of cpu per of your data pre-processing instance   
+`iterations` no. of iterations in an epoch - must be multiple of 10s`   
+`grpc-workers` no. of workers fetching the pre-processed data to DNN process (gRPC client)   
+
+**Step 2**: Run gRPC Client in a new terminal session (File > New > Terminal) . The set of process spawend by this script fetches pre-processed data from server and runs training. Make sure you change your working directory to where code exist (..\..\pt.grpc.local). 
+```
+cd ~/SageMaker/hetro-training/pt.grpc.local/
+python main_grpc_client.py --batch-size 8192 --num-dnn-workers 2 --iterations 100 --model-dir ./
+```
+where,   
+`batch-size` any integer, must match to the size mentioned in the gRPC server process launch   
+`num-workers` no. of dataloader workers, it is based on no. of cpu of the dnn instance   
+`iterations` no. of iterations in an epoch - must be multiple of 10s, must match to the no. mentioned in the gRPC server process launch   
+`model-dir` location of the model to be saved   
+
+**Step 3**: Observe the `avg step time`. The steps/second starts printing to the console. And, stops after predefined iterations.
+```    
+sh-4.2$ python main_grpc_client.py 
+Training job started...
+10: avg step time: 0.43338242229997376
+20: avg step time: 0.3908786807500064
+30: avg step time: 0.3767881167999955
+40: avg step time: 0.3698345946000018
+50: avg step time: 0.3657775266399949
+60: avg step time: 0.5407461890666657
+70: avg step time: 0.6490721581571441
+80: avg step time: 0.7544923407874989
+90: avg step time: 0.8115056776666633
+100: avg step time: 0.8806567812599997
+Saving the model
+Training job completed!
+Shutting down data service via port 16000
+``` 
+    
+**Step 4**: Optionally, in a new terminal window, you can validate whether the gRPC client-server communication is taking place. 
```

Comment:
There are security implications throughout (lots of network communications and popping shells), but no security section in the documentation. This should be addressed.